flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.java

```diff
@@ -0,0 +1,221 @@
+		long timestamp,
+		KeyedProcessFunction<K, BaseRow, BaseRow>.OnTimerContext ctx,
+		Collector<BaseRow> out) throws Exception {
+		if (needToCleanupState(timestamp)) {
+			// clean up and return
+			cleanupState(inputState, accState);
+			function.cleanup();
+			return;
+		}
+
+		// remove timestamp set outside of ProcessFunction.
+		((TimestampedCollector) out).eraseTimestamp();
+
+		// we consider the original timestamp of events
+		// that have registered this time trigger 1 ms ago
+
+		long currentTime = timestamp - 1;
+		// get the list of elements of current proctime
+		List<BaseRow> currentElements = inputState.get(currentTime);
+
+		// Expired clean-up timers pass the needToCleanupState check.
+		// Perform a null check to verify that we have data to process.
+		if (null == currentElements) {
+			return;
+		}
+
+		// initialize the accumulators
+		BaseRow accumulators = accState.value();
+
+		if (null == accumulators) {
+			accumulators = function.createAccumulators();
+		}
+
+//		// set accumulators in context first
+//		function.setAccumulators(accumulators);
+
+		// update the elements to be removed and retract them from aggregators
+		long limit = currentTime - precedingTimeBoundary;
+
+		// we iterate through all elements in the window buffer based on timestamp keys
+		// when we find timestamps that are out of interest, we retrieve corresponding elements
+		// and eliminate them. Multiple elements could have been received at the same timestamp
+		// the removal of old elements happens only once per proctime as onTimer is called only once
+		Iterator<Map.Entry<Long, List<BaseRow>>> iter = inputState.iterator();
+		while (iter.hasNext()) {
+			Map.Entry<Long, List<BaseRow>> entry = iter.next();
+			Long elementKey = entry.getKey();
+			if (elementKey < limit) {
+				// element key outside of window. Retract values
+				List<BaseRow> elementsRemove = entry.getValue();
+				int iRemove = 0;
+				while (iRemove < elementsRemove.size()) {
+					BaseRow retractRow = elementsRemove.get(iRemove);
+					function.retract(retractRow);
+					iRemove += 1;
+				}
+				iter.remove();
+			}
+		}
+
+		// set accumulators in context first
+		function.setAccumulators(accumulators);
+
+		// add current elements to aggregator. Multiple elements might
+		// have arrived in the same proctime
+		// the same accumulator value will be computed for all elements
+		int iElemenets = 0;
+		while (iElemenets < currentElements.size()) {
+			BaseRow input = currentElements.get(iElemenets);
+			function.accumulate(input);
+			iElemenets += 1;
+		}
+
+		// we need to build the output and emit for every event received at this proctime
+		iElemenets = 0;
+		BaseRow aggValue = function.getValue();
+		while (iElemenets < currentElements.size()) {
+			BaseRow input = currentElements.get(iElemenets);
+
+			output.replace(input, aggValue);
+			out.collect(output);
+			iElemenets += 1;
+		}
+
+		// update the value of accumulators for future incremental computation
```

Comment:
We need to retrieve the accumulator because it might be modified in the previous while iteration.

```scala
accumulators = function.getAccumulators
```